OpenRA.Platforms.Default/OpenAlSoundEngine.cs

```diff
@@ -483,196 +491,105 @@ public virtual void Stop()
 		readonly CancellationTokenSource cts = new CancellationTokenSource();
-		readonly Task streamTask;
-		readonly Stack<uint> freeBuffers = new Stack<uint>(BufferCount);
-		int totalSamplesPlayed;
+		readonly Task playTask;
 
-		public OpenAlStreamingSound(uint source, bool looping, bool relative, WPos pos, float volume, int channels, int sampleBits, int sampleRate, Stream stream)
+		public OpenAlAsyncLoadSound(uint source, bool looping, bool relative, WPos pos, float volume, int channels, int sampleBits, int sampleRate, Stream stream)
 			: base(source, looping, relative, pos, volume, sampleRate)
 		{
-			streamTask = Task.Run(async () =>
-			{
-				var format = OpenAlSoundEngine.MakeALFormat(channels, sampleBits);
-				var bytesPerSample = sampleBits / 8;
+			// Load a silent buffer into the source. Without this,
+			// attempting to change the state (i.e. play/pause) the source fails on some systems.
+			var silentSource = new OpenAlSoundSource(SilentData, channels, sampleBits, sampleRate);
+			AL10.alSourcei(source, AL10.AL_BUFFER, (int)silentSource.Buffer);
 
-				var buffers = new uint[BufferCount];
-				AL10.alGenBuffers(new IntPtr(buffers.Length), buffers);
-				try
+			playTask = Task.Run(async () =>
+			{
+				MemoryStream memoryStream;
+				using (stream)
 				{
-					foreach (var buffer in buffers)
-						freeBuffers.Push(buffer);
-					var data = new byte[sampleRate * bytesPerSample * BufferSizeInSecs];
-					var streamEnd = false;
-
-					while (!streamEnd && !cts.IsCancellationRequested)
+					memoryStream = new MemoryStream();
+					try
 					{
-						// Fill the data array as fully as possible.
-						var count = await ReadFillingBuffer(stream, data);
-						streamEnd = count < data.Length;
+						await stream.CopyToAsync(memoryStream, 81920, cts.Token);
+					}
+					catch (TaskCanceledException)
+					{
+						// Sound was stopped early, cleanup the unused buffer and exit.
+						AL10.alSourceStop(source);
+						AL10.alSourcei(source, AL10.AL_BUFFER, 0);
+						silentSource.Dispose();
+						return;
+					}
+				}
 
-						// Fill a buffer and queue it for playback.
-						var nextBuffer = freeBuffers.Pop();
-						AL10.alBufferData(nextBuffer, format, data, new IntPtr(count), new IntPtr(sampleRate));
-						AL10.alSourceQueueBuffers(source, new IntPtr(1), ref nextBuffer);
+				var data = memoryStream.ToArray();
+				var bytesPerSample = sampleBits / 8f;
+				var lengthInSecs = data.Length / bytesPerSample / sampleRate;
+				using (var soundSource = new OpenAlSoundSource(data, channels, sampleBits, sampleRate))
+				{
+					// Need to stop the source, before attaching the real input and deleting the silent one.
+					AL10.alSourceStop(source);
+					AL10.alSourcei(source, AL10.AL_BUFFER, (int)soundSource.Buffer);
+					silentSource.Dispose();
 
-						lock (cts)
+					{
+						int state;
+						AL10.alGetSourcei(Source, AL10.AL_SOURCE_STATE, out state);
+						if (state != AL10.AL_STOPPED)
```

Comment:
We need to be careful about race conditions here.  If someone triggers a pause from the main thread between these two calls then we may end up playing instead of rewinding (or vice versa).  The consequences are relatively minor, so this probably doesn't need to block merging.

Reply:
But a `// TODO: ` acknowledging the potential issue is probably worthwhile.